### ISSUE TITLE
After model compilation and run on second gpu card (cuda:1) , 154 MB still goes to the first gpu card (cuda:0).Using pytorch ignite framework

### ISSUE THREAD

**Issue by nomaad42** [pytorch/ignite#3187]
## 🐛 Bug description

EDIT: After model compilation and run on the second GPU card (cuda:1), 154 MB still goes to the first GPU card (cuda:0). Using the PyTorch ignite framework

Without compilation --> it's not a problem

![image](https://github.com/pytorch/ignite/assets/62198792/cae4e46a-e88a-4903-b967-bad9f4384718)

## Environment

- PyTorch Version (e.g., 1.4): 2.1.2
- Ignite Version: 0.4.13
- OS (e.g., Linux): Ubuntu 20.04
- How you installed Ignite: pip
- Python version: 3.10.12

**Comment by vfdev-5** [comment]
@nomaad42 thanks for reporting, can you share some code to repro the problem ?

**Comment by nomaad42** [comment]
@vfdev-5 

Sure, is this enough?

`model = swin_v2_t(weights=Swin_V2_T_Weights.IMAGENET1K_V1)`
`model.head = Linear(model.head.in_features, 2)`
`compiled_model = torch.compile(model)`
`compiled_model.to(device)`

`trainer = create_supervised_trainer(model, optimizer, criterion, device=device)`

`trainer.run(train_loader, max_epochs=epochs)`

**Comment by vfdev-5** [comment]
And `device` is `cuda:1` ?

**Comment by nomaad42** [comment]
Exactly

**Comment by vfdev-5** [comment]
After checking in details your repro code, I would say it is not pytorch-ignite issue but pytorch core. Here is a complete repro code snippet without using ignite:
```python
import torch
from torch.nn import Linear
from torchvision.models import swin_v2_t, Swin_V2_T_Weights

print("CUDA Devices:")
for d in [0, 1]:
    print("\t- " + torch.cuda.get_device_name(device=d))


def print_memory_usage():
    for d in [0, 1]:
        stats = torch.cuda.memory_stats(device=d)
        m = stats["allocated_bytes.all.allocated"] + stats["inactive_split_bytes.all.allocated"] + stats["reserved_bytes.all.allocated"]
        print(f"\t- CUDA Device: {d}, allocated + reserved + non-released in GB: {m / 1024 / 1024}")


device = "cuda:1"
model = swin_v2_t(weights=Swin_V2_T_Weights.IMAGENET1K_V1)
model.head = Linear(model.head.in_features, 2)

compiled_model = torch.compile(model)

print("--- Before compiled model to device")
print_memory_usage()

compiled_model.to(device)
x = torch.rand(16, 3, 320, 320, device=device)

print("--- Before compiled model forward")
print_memory_usage()

y = compiled_model(x)

print("--- Before compiled model backward")
print_memory_usage()

y.sum().backward()

print("--- After compiled model backward")
print_memory_usage()
```
It gives the following output:
```
CUDA Devices:
        - Quadro RTX 8000
        - Quadro RTX 8000
--- Before compiled model to device
        - CUDA Device: 0, allocated + reserved + non-released in GB: 0.0
        - CUDA Device: 1, allocated + reserved + non-released in GB: 0.0
--- Before compiled model forward
        - CUDA Device: 0, allocated + reserved + non-released in GB: 0.0
        - CUDA Device: 1, allocated + reserved + non-released in GB: 371.6181640625
--- Before compiled model backward
        - CUDA Device: 0, allocated + reserved + non-released in GB: 8.021484375
        - CUDA Device: 1, allocated + reserved + non-released in GB: 16637.2783203125
--- After compiled model backward
        - CUDA Device: 0, allocated + reserved + non-released in GB: 8.021484375
        - CUDA Device: 1, allocated + reserved + non-released in GB: 19492.46923828125
```


So, I confirm that during the torch compilation something is using internally device=0 that you also see with `nvidia-smi`. I can report the issue on pytorch for further investigations.
Thanks again for reporting !

**Comment by nomaad42** [comment]
@vfdev-5 , Thank you very much for your prompt investigation!
